src/nvim/api/vim.c

```diff
@@ -164,6 +164,83 @@ Object vim_eval(String str, Error *err)
   return rv;
 }
 
+/// Call the given function with the given arguments stored in an array.
+///
+/// @param functionName The function to call
+/// @param args the functions arguments packed in an Array
+/// @param self optional self dictionary that may be used by user functions to
+///        emulate object-oriented programming
+/// @param[out] err Details of an error that may have occurred
+/// @return The result of the function call if it was successful
+Object vim_call_function(String functionName, Array args, Object self,
+                         Error *err)
+  FUNC_ATTR_DEFERRED
+{
+  Object rv = OBJECT_INIT;
+  if (args.size > MAX_FUNC_ARGS) {
+    api_set_error(err, Validation,
+      _("Function called with too many arguments."));
+    return rv;
+  }
+
+  typval_T optional_self_dict;
+  dict_T *self_dict = NULL;
+  if (self.type == kObjectTypeNil) {
+    self_dict = NULL;
+  } else if (self.type == kObjectTypeDictionary
+             || self.type == kObjectTypeArray) {
+    // XXX The tests somehow cause the object to be an array
+    self.type = kObjectTypeDictionary;
+    if (!object_to_vim(self, &optional_self_dict, err)) {
+      free_tv(&optional_self_dict);
+      return rv;
+    } else if (optional_self_dict.v_type == VAR_DICT) {
+      self_dict = optional_self_dict.vval.v_dict;
+    } else {
+      api_set_error(err, Validation,
+          _("Object was not converted to a dict."));
+      goto free_vim_args;
+    }
+  } else {
```

Comment:
Since I mostly pass empty dictionary here I do not care much. But for running things like `g:SyntasticLoclist.current().filter(…)` this may be handy. I would though say that in this case string must be simply `eval`ed, not that it should be a name of global dictionary.

About running global objects: it may make sense to first check functionName for presense in self dictionary and only then pass it to call_func untouched.

Or, maybe a better idea: strip off `self` handling from `vim_call_function` completely (restore the original functionality), but additionally add `vim_call_dict_function(Object self, Boolean internal, String functionName, Array args, Error *err)` which will do the following:

1. Convert `self` to a dictionary: it should either be a Dictionary or a String object, in first case it is simply converted dictionary, in the last case it is expression passed to `eval()` which must evaluate to a dictionary.
2. If `internal` is true then check whether `functionName` is a key in `self` that refers to a function reference. If yes, replace function name with value of this reference. If no, raise an error. If `internal` is false then `functionName` is passed to `call_func` untouched.
3. Call `call_func` with arguments and self dictionary.

Intended usage: same as `vim_call_func`, but for constructs like `g:SyntasticLoclist.current().filter(…)`: `g:SyntasticLoclist.current()` goes into `self` as a string, arguments of `.filter()` may be different and go into `args`, rationale is the same: first string is a simple static string that does not change, arguments in this case do not require double serializing (from dictionary expected by `.filter()` to string then string to messagepack).

Additional rationale (behind the `internal` argument): this make it possible to call anonymous functions without knowing their internal name (i.e. number), thus anonymous functions can be truly anonymous like in extended-funcref branch (they are not stored in a global functions dictionary under any name there). (Or in VimL-to-lua translator, if I remember correctly.)